controllers/agents/app/servicebinding_controller.go

```diff
@@ -47,13 +60,326 @@ type ServiceBindingReconciler struct {
+	if err != nil {
+		l.Error(err, "unable to convert volumeProjection to an unstructured object")
+		return ctrl.Result{}, err
+	}
+	applications, result, err := r.getApplication(ctx, req, serviceBinding, secretName)
+	if err != nil {
+		return result, err
+	}
+	secretLookupKey := client.ObjectKey{Name: serviceBinding.Spec.ServiceEndpointDefinitionSecret, Namespace: req.NamespacedName.Namespace}
+	psSecret := &v1.Secret{}
+	if err := r.Get(ctx, secretLookupKey, psSecret); err != nil {
+		return ctrl.Result{}, err
+	}
+	return r.bindApplications(ctx, req, serviceBinding, psSecret, mountPathDir, volumeName, unstructuredVolume, applications...)
+}
+
+func (r *ServiceBindingReconciler) bindApplications(ctx context.Context, req ctrl.Request,
+	sb primazaiov1alpha1.ServiceBinding, psSecret *v1.Secret, mountPathDir, volumeName string, unstructuredVolume map[string]interface{}, applications ...unstructured.Unstructured) (ctrl.Result, error) {
+
+	l := log.FromContext(ctx)
+
+	var el error
+	for index, application := range applications {
+		containersPaths := [][]string{}
+		volumesPath := []string{"spec", "template", "spec", "volumes"}
+
+		containersPaths = append(containersPaths,
+			[]string{"spec", "template", "spec", "containers"},
+			[]string{"spec", "template", "spec", "initContainers"},
+		)
+		l.Info("referencing the volume in an unstructured object")
+		volumes, found, err := unstructured.NestedSlice(application.Object, volumesPath...)
+		if err != nil {
+			l.Error(err, "unable to reference the volumes in the application object")
+			return ctrl.Result{}, err
+		}
+		if !found {
+			l.Info("volumes not found in the application object")
+		}
+		l.Info("Volumes values", "volumes", volumes)
+
+		volumeFound := false
+
+		for i, volume := range volumes {
+			l.Info("Volume", "volume", volume)
+			if volume.(map[string]interface{})["name"].(string) == volumeName {
+				volumes[i] = unstructuredVolume
+				volumeFound = true
+				break
+			}
+		}
+
+		if !volumeFound {
+			volumes = append(volumes, unstructuredVolume)
+		}
+		l.Info("setting the updated volumes into the application using the unstructured object")
+		if err := unstructured.SetNestedSlice(application.Object, volumes, volumesPath...); err != nil {
+			return ctrl.Result{}, err
+		}
+		l.Info("application object after setting the update volume", "Application", application)
+
+		for _, containersPath := range containersPaths {
+			l.Info("referencing containers in an unstructured object")
+			containers, found, err := unstructured.NestedSlice(application.Object, containersPath...)
+			if err != nil {
+				l.Error(err, "unable to reference containers in the application object")
+				return ctrl.Result{}, err
+			}
+			if !found {
+				e := &field.Error{Type: field.ErrorTypeRequired, Field: strings.Join(containersPath, "."), Detail: "no containers"}
+				l.Info("containers not found in the application object", "error", e)
+			}
+
+		CONTAINERS_OUTER:
```

Comment:
If you like it, instead of using the label you may move this code in a new function

Reply:
Yeah let's do that